src/libcore/iter/mod.rs

```diff
@@ -1099,6 +1099,16 @@ impl<I: Iterator, P> Iterator for Filter<I, P> where P: FnMut(&I::Item) -> bool
         let (_, upper) = self.iter.size_hint();
         (0, upper) // can't know a lower bound, due to the predicate
     }
+
+    #[inline]
+    fn count(self) -> usize {
+        let (mut c, mut predicate, mut iter) = (0, self.predicate, self.iter);
+        for x in iter.by_ref() {
+            // branchless count
+            c += (&mut predicate)(&x) as usize;
+        }
+        c
```

Comment:
Maybe one doesn't want to argue style for trivial things, but I find the code overly complicated when simple should work well:

```rust
fn count(mut self) -> usize {
    // Attempt to produce branchless count if possible
    let mut count = 0;
    for elt in self.iter {
        count += (self.predicate)(&elt) as usize;
    }
    count
}
```

The surrounding methods use `.by_ref()` which seems archaic (just `&mut self.iter` would be fine if mutable reference only was required)